Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "RemoteLayerTreeOverlayRegionLayers.h"
+
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+
+#import "PlatformCALayerRemote.h"
+#import "RemoteLayerTreeHost.h"
+#import "WKScrollView.h"
+#import "WKWebViewIOS.h"
+#import <QuartzCore/QuartzCore.h>
+#import <WebCore/IntRectHash.h>
+#import <objc/runtime.h>
+#import <pal/spi/cocoa/QuartzCoreSPI.h>
+
+@interface UIView (UIViewInternal_RemoteLayerTreeOverlayRegionLayers)
+- (UIViewController *)_viewControllerForAncestor;
+@end
+
+static inline NSData* dataFromRect(CGRect rect)
+{
+    return [NSData dataWithBytes:(void *)&rect length:sizeof(CGRect)];
+}
+
+@interface WKScrollView (RemoteLayerTreeOverlayRegionLayersAdditions)
+@property (nonatomic, readonly) NSArray<NSData *> *overlayRegions;
+@property (nonatomic, readonly) NSMapTable<UIView *, NSData *> *internalOverlayRegions;
+@end
+
+@implementation WKScrollView (RemoteLayerTreeOverlayRegionLayersAdditions)
+- (NSArray<NSData *> *)overlayRegions
+{
+    return NSAllMapTableValues(self.internalOverlayRegions);
+}
+
+- (NSMapTable<UIView *, NSData *> *)internalOverlayRegions
+{
+    static void* internalOverlayRegionsKey = &internalOverlayRegionsKey;
+
+    NSMapTable *mapTable = objc_getAssociatedObject(self, internalOverlayRegionsKey);
```

Comment:
Since WKScrollView is *ours*, you could just add an internal property instead of using associated objects here, no?

Reply:
I would prefer this WKScrollView code to exist in WKScrollView.mm

Reply:
Sounds good, that is easier than using associated objects too